sys/include/event/bus.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2022 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_event
+ * @brief       Provides functionality to trigger multiple events at once
+ *
+ * @{
+ *
+ * @file
+ * @brief       Event Bus API
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ */
+
+#ifndef EVENT_BUS_H
+#define EVENT_BUS_H
+
+#include "event.h"
+#include "list.h"
+#include "mutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Event Bus struct
+ */
+typedef struct {
+    list_node_t list;       /**< event bus list node                   */
+    event_queue_t *queue;   /**< event queue to be used for bus events */
+    mutex_t lock;           /**< bus mutex                             */
+} event_bus_t;
+
+/**
+ * @brief   Entry on the Event Bus
+ */
+typedef struct {
+    list_node_t node;       /**< event bus list node  */
+    event_t *event;         /**< pointer to bus event */
+} event_bus_entry_t;
+
+/**
+ * @brief   Initialize an event bus
+ *
+ * @param[in]   q       The event queue to use for the bus
+ */
+#define EVENT_BUS_INIT(q)   { .queue = q }
+
+/**
+ * @brief   Initialize a bus event
+ *
+ * @param[in]   e       The event to use for the bus event
+ */
+#define EVENT_BUS_ENTRY_INIT(e) { .event = (event_t *)e }
+
+/**
+ * @brief   Add an event to the event bus
+ *
+ * @param[in]   bus     The event bus to add the event to
+ * @param[in]   event   The new bus event
+ */
+static inline void event_bus_add(event_bus_t *bus, event_bus_entry_t *event)
+{
+    mutex_lock(&bus->lock);
+
+    list_add(&bus->list, &event->node);
+
+    mutex_unlock(&bus->lock);
+}
+
+/**
+ * @brief   Remove an event to the event bus
+ *
+ * @param[in]   bus     The event bus to remove the event from
+ * @param[in]   event   The bus event to remove
+ */
+static inline void event_bus_remove(event_bus_t *bus, event_bus_entry_t *event)
+{
+    mutex_lock(&bus->lock);
+
+    list_remove(&bus->list, &event->node);
+
+    mutex_unlock(&bus->lock);
+}
+
+/**
+ * @brief   Trigger all events registered on the bus
+ *
+ * @param[in]   bus     The event bus to trigger
+ */
+static inline void event_bus_post(event_bus_t *bus)
+{
+    mutex_lock(&bus->lock);
```

Comment:
Using a mutex here makes it impossible to pose from ISR.

Also, the first post might context switch right away, then the next, ..., depending on the caller's (and handler thread's) priority.
Using a critical section would make this IRQ safe, and allow a low-prio thread to trigger multiple events and let the scheduler sort it out. That would also effectively avoid a triggered thread to cause a dead-lock through modifying the bus (e.g., removing itself) while the handler is running.

Reply:
(yeah, at the cost of another irq-dead time while traversing a list)

Reply:
I think in the case a low-prio thread posts the bus, using a critical section would save a lot of cycles, as the list get's traversed once, then the scheduler would switch once to each thread (with per-entry queue) or to the handler thread) in regular thread priority.
When using a mutex, every `event_post()` would cause a single context switch, then a single event handled, then another context switch back to the posting thread, ..., so one extra context switch per event.

Reply:
> (yeah, at the cost of another irq-dead time while traversing a list)

Probably, each extra context switch blocks IRQs longer than it takes to traverse the list in a critical section, for reasonable list lengths. E.g., on ARM, a context switch is >250cycles.